### PR TITLE
Support PHP ITs

### PR DESCRIPTION
**Pull request by v1v**
## What does this PR do?

- Enable PHP in the ITs
- Enable PHP in the test-selector-ITs
- Support PHP versions and build from source code.

```
./scripts/compose.py list-options | grep php 
--agent-php-apache-port
--no-agent-php-apache
--php-agent-release
--php-agent-repo
--php-agent-version
--with-agent-php-apache
```

## Why is it important?

Integrate APM-ITs in the same manner we have done for the other APM projects.

## Related issues

Required by https://github.com/elastic/apm-agent-php/pull/236
Depends on https://github.com/elastic/apm-pipeline-library/pull/778

## Tests

- With default parameters then it will compile the php agent from source code

<details><summary>Expand to view the logs</summary>
<p>

```bash
$ .ci/scripts/php.sh
...
Step 8/8 : RUN ./run.sh
 ---> Running in c851d4f8441f
+ '[' -z '' ']'
+ git clone https://github.com/elastic/apm-agent-php.git /src/apm-agent-php
Cloning into '/src/apm-agent-php'...
+ cd /src/apm-agent-php
+ git fetch -q origin '+refs/pull/*:refs/remotes/origin/pr/*'
+ git checkout master
Already on 'master'
Your branch is up to date with 'origin/master'.
...
Installing shared extensions:     /usr/local/lib/php/extensions/no-debug-non-zts-20190902/
+ cp /src/ext-elastic-apm.ini /usr/local/etc/php/conf.d/ext-elastic-apm.ini
+ cp -rf /src/app /var/www/html/
Removing intermediate container c851d4f8441f
 ---> cdc615e1b6b3
...
```

</p>
</details>

- With release parameter then it will download and install the release for the php agent.

<details><summary>Expand to view the logs</summary>
<p>

```bash
$ APM_AGENT_PHP_VERSION="release;0.2" .ci/scripts/php.sh
...
 ---> Running in b9a586c8e688
+ '[' -z 0.2 ']'
+ wget -q https://github.com/elastic/apm-agent-php/releases/download/v0.2/apm-agent-php_0.2_all.deb -O /tmp/apm-agent-php.deb
+ dpkg -i /tmp/apm-agent-php.deb
Selecting previously unselected package apm-agent-php.
(Reading database ... 14578 files and directories currently installed.)
Preparing to unpack /tmp/apm-agent-php.deb ...
Unpacking apm-agent-php (0.2) ...
Setting up apm-agent-php (0.2) ...
Installing Elastic PHP agent
/usr/local/etc/php/php.ini has been created with the Elastic PHP agent setup.
; THIS IS AN AUTO-GENERATED FILE by the Elastic PHP agent post-install.sh script
extension=/opt/elastic/apm-agent-php/extensions/elastic_apm-20190902.so
elastic_apm.bootstrap_php_part_file=/opt/elastic/apm-agent-php/src/bootstrap_php_part.php
; END OF AUTO-GENERATED
Extension enabled successfully for Elastic PHP agent
+ cp -rf /src/app /var/www/html/
Removing intermediate container b9a586c8e688
 ---> f24253b9093b

```

</p>
</details>

- Using the python compose.py script for a given sha commit -> https://github.com/elastic/apm-agent-php/commit/e3bc93f7d7dd70b6f15e9cd0aab2494b30d6953f

<details><summary>Expand to view the logs</summary>
<p>

```bash
$ python3 scripts/compose.py start 8.0.0  \
      --build-parallel \
      --php-agent-version=e3bc93f7d7dd70b6f15e9cd0aab2494b30d6953f \
      --no-apm-server-dashboards \
      --no-apm-server-self-instrument \
      --with-agent-php-apache \
      --apm-server-agent-config-poll=1s \
      --force-build \
      --no-xpack-secure
...
 ---> Running in a4ae87b9700b
+ '[' -z '' ']'
+ git clone https://github.com/elastic/apm-agent-php.git /src/apm-agent-php
Cloning into '/src/apm-agent-php'...
+ cd /src/apm-agent-php
+ git fetch -q origin '+refs/pull/*:refs/remotes/origin/pr/*'
+ git checkout e3bc93f7d7dd70b6f15e9cd0aab2494b30d6953f
Note: checking out 'e3bc93f7d7dd70b6f15e9cd0aab2494b30d6953f'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at e3bc93f Disable APM on the process running the component tests (#230)
+ git rev-parse HEAD
e3bc93f7d7dd70b6f15e9cd0aab2494b30d6953f
+ cd /src/apm-agent-php/src/ext
+ phpize
Configuring for:
PHP Api Version:         20190902
Zend Module Api No:      20190902
Zend Extension Api No:   320190902
+ ./configure --enable-elastic_apm
...
Installing shared extensions:     /usr/local/lib/php/extensions/no-debug-non-zts-20190902/
+ cp /src/ext-elastic-apm.ini /usr/local/etc/php/conf.d/ext-elastic-apm.ini
+ cp -rf /src/app/foo /src/app/healthcheck /src/app/info.php /var/www/html
Removing intermediate container a4ae87b9700b
 ---> 8e7a9642f371

Successfully built 8e7a9642f371
Building agent-php-apache ... done
Pulling elasticsearch ... done
Pulling kibana        ... done
Pulling apm-server    ... done
localtesting_8.0.0_elasticsearch is up-to-date
localtesting_8.0.0_kibana is up-to-date
localtesting_8.0.0_apm-server is up-to-date
Recreating phpapacheapp ... done
$ echo $?
0
$ docker ps                                                               
CONTAINER ID        IMAGE                                                          COMMAND                  CREATED             STATUS                    PORTS                                                                                                        NAMES
29374dad27e9        apm-integration-testing_agent-php-apache                       "docker-php-entrypoi…"   54 seconds ago      Up 53 seconds (healthy)   127.0.0.1:8030->80/tcp                                                                                       phpapacheapp
b0ffb3c27da4        docker.elastic.co/apm/apm-server:8.0.0-SNAPSHOT                "/usr/local/bin/dock…"   5 minutes ago       Up 5 minutes (healthy)    127.0.0.1:6060->6060/tcp, 127.0.0.1:8200->8200/tcp, 127.0.0.1:14250->14250/tcp, 127.0.0.1:14268->14268/tcp   localtesting_8.0.0_apm-server
bb19ab2199f3        docker.elastic.co/kibana/kibana:8.0.0-SNAPSHOT                 "/usr/local/bin/dumb…"   6 minutes ago       Up 6 minutes (healthy)    127.0.0.1:5601->5601/tcp                                                                                     localtesting_8.0.0_kibana
37404a58cd55        docker.elastic.co/elasticsearch/elasticsearch:8.0.0-SNAPSHOT   "/bin/tini -- /usr/l…"   6 minutes ago       Up 6 minutes (healthy)    127.0.0.1:9200->9200/tcp, 9300/tcp                                                                           localtesting_8.0.0_elasticsearch

```

</p>
</details>

- Using the python compose.py script for a given release

<details><summary>Expand to view the logs</summary>
<p>

```bash
$ python3 scripts/compose.py start 8.0.0  \
      --build-parallel \
      --php-agent-release=0.2 \
      --no-apm-server-dashboards \
      --no-apm-server-self-instrument \
      --with-agent-php-apache \
      --apm-server-agent-config-poll=1s \
      --force-build \
      --no-xpack-secure
...
 ---> Running in 49b00aae72c8
+ '[' -z 0.2 ']'
+ wget -q https://github.com/elastic/apm-agent-php/releases/download/v0.2/apm-agent-php_0.2_all.deb -O /tmp/apm-agent-php.deb
+ dpkg -i /tmp/apm-agent-php.deb
Selecting previously unselected package apm-agent-php.
(Reading database ... 14578 files and directories currently installed.)
Preparing to unpack /tmp/apm-agent-php.deb ...
Unpacking apm-agent-php (0.2) ...
Setting up apm-agent-php (0.2) ...
Installing Elastic PHP agent
/usr/local/etc/php/php.ini has been created with the Elastic PHP agent setup.
; THIS IS AN AUTO-GENERATED FILE by the Elastic PHP agent post-install.sh script
extension=/opt/elastic/apm-agent-php/extensions/elastic_apm-20190902.so
elastic_apm.bootstrap_php_part_file=/opt/elastic/apm-agent-php/src/bootstrap_php_part.php
; END OF AUTO-GENERATED
Extension enabled successfully for Elastic PHP agent
+ cp -rf /src/app/foo /src/app/healthcheck /src/app/info.php /var/www/html
Removing intermediate container 49b00aae72c8
 ---> 722dc4fdf928

Successfully built 722dc4fdf928
Building agent-php-apache ... done
Pulling elasticsearch ... done
Pulling kibana        ... done
Pulling apm-server    ... done
Creating network "apm-integration-testing" with the default driver
Creating volume "apm-integration-testing_esdata" with local driver
Creating volume "apm-integration-testing_pgdata" with local driver
Creating localtesting_8.0.0_elasticsearch ... done
Creating localtesting_8.0.0_kibana        ... done
Creating localtesting_8.0.0_apm-server    ... done
Creating phpapacheapp                     ... done
$ echo $?                                                                         
0
$ docker ps
CONTAINER ID        IMAGE                                                          COMMAND                  CREATED              STATUS                        PORTS                                                                                                        NAMES
b50f052e34c6        apm-integration-testing_agent-php-apache                       "docker-php-entrypoi…"   22 seconds ago       Up 21 seconds (healthy)       127.0.0.1:8030->80/tcp                                                                                       phpapacheapp
b0ffb3c27da4        docker.elastic.co/apm/apm-server:8.0.0-SNAPSHOT                "/usr/local/bin/dock…"   33 seconds ago       Up 32 seconds (healthy)       127.0.0.1:6060->6060/tcp, 127.0.0.1:8200->8200/tcp, 127.0.0.1:14250->14250/tcp, 127.0.0.1:14268->14268/tcp   localtesting_8.0.0_apm-server
bb19ab2199f3        docker.elastic.co/kibana/kibana:8.0.0-SNAPSHOT                 "/usr/local/bin/dumb…"   43 seconds ago       Up 43 seconds (healthy)       127.0.0.1:5601->5601/tcp                                                                                     localtesting_8.0.0_kibana
37404a58cd55        docker.elastic.co/elasticsearch/elasticsearch:8.0.0-SNAPSHOT   "/bin/tini -- /usr/l…"   About a minute ago   Up About a minute (healthy)   127.0.0.1:9200->9200/tcp, 9300/tcp     
```

</p>
</details>

### CODE REVIEW

**Review by mdelapenya — APPROVED** [review]

**Review by graphaelli — APPROVED** [review]

**Comment by graphaelli** [comment]
New failure here, @SergeyKleyman is this because of https://github.com/elastic/apm-agent-php/pull/235?

```
[2020-10-19T17:40:44.178Z]                 if 'stacktrace' in span.keys():
[2020-10-19T17:40:44.178Z]                     stacktrace = span['stacktrace']
[2020-10-19T17:40:44.178Z]                     assert 15 < len(stacktrace) < 70, \
[2020-10-19T17:40:44.178Z]                         "number of frames not expected, got {}, but this assertion might be too strict".format(
[2020-10-19T17:40:44.178Z] >                           len(stacktrace))
[2020-10-19T17:40:44.178Z] E                   AssertionError: number of frames not expected, got 1, but this assertion might be too strict

```

**Comment by SergeyKleyman** [comment]
> New failure here, @SergeyKleyman is this because of [elastic/apm-agent-php#235](https://github.com/elastic/apm-agent-php/pull/235)?

I'll take a look at it ASAP. I don't think it's related to [elastic/apm-agent-php#235](https://github.com/elastic/apm-agent-php/pull/235) - that change was to address technical debt (inconsistency between two methods in agent's public API). My current hypothesis that I am going to investigate first is that [$span->end();](https://github.com/elastic/apm-integration-testing/blob/bc742199a2c598905328e653a1387c20dd5a77f2/docker/php/apache/app/foo/index.php#L12) is called at the top level of the script so the stack is empty. I didn't test such use case.

**Review by cachedout — APPROVED** [review]

**Review by kuisathaverat — APPROVED** [review]

File: .ci/integrationTestSelector.groovy [hunk 1]
```diff
@@ -90,7 +90,8 @@ pipeline {
           agent { label 'linux && immutable' }
           options { skipDefaultCheckout() }
           when {
-            expression { return (params.INTEGRATION_TEST != 'RUM') }
+            expression { return (params.INTEGRATION_TEST != 'RUM' &&
+                                 params.INTEGRATION_TEST != 'PHP') }  // TODO: When Opbeans PHP app is available then enable it
```
Comment (v1v) [inline .ci/integrationTestSelector.groovy:94]: NOTE: As soon as there is a PHP Opbeans project then we will be able to enable the validation

**Comment by v1v** [comment]
I'll proceed to merge this PR since test failures are already happening in the master branch anyway

**Comment by v1v** [comment]
@Mergifyio backport 7.x

**Pull request merged**